modules/db_postgres/pg_con.c

```diff
@@ -19,87 +19,169 @@
+
+			if (ptr->tls_dom->verify_cert == 1) {
+				PSQL_PARAM("sslmode", "verify-ca");
+			}
+
+			PSQL_PARAM("sslkey", ptr->tls_dom->pkey.s)
+			PSQL_PARAM("sslcert", ptr->tls_dom->cert.s)
+			PSQL_PARAM("sslrootcert", ptr->tls_dom->ca.s)
+		}
 	}
-	LM_DBG("%p=pkg_malloc(%zu)\n", ptr, sizeof(struct pg_con));
 
-	memset(ptr, 0, sizeof(struct pg_con));
-	ptr->ref = 1;
+	/* force the default timeout */
+	if (pq_timeout > 0) {
+		PSQL_PARAM("connect_timeout", int2str(pq_timeout, 0));
+	}
 
 	if (id->parameters) {
-		lend = strlen(id->database);
+
 		lenp = strlen(id->parameters);
-		dbname = pkg_malloc(7 /* "dbname=" */ +
-				lend + 1 /* ? */ + lenp + 1 /* '\0' */);
+		LM_DBG("id->parameters(%d): %s\n", lenp, id->parameters);
+
+		dbname = pkg_malloc(lenp + 1 /* '\0' */);
+		memset(dbname, 0, lenp + 1);
+
 		if (!dbname) {
-			LM_ERR("oom for building database name!\n");
-			goto err;
+			LM_ERR("oom for building database connection string!\n");
+			return -1;
 		}
-		memcpy(dbname, "dbname=", 7);
-		memcpy(dbname + 7, id->database, lend);
-		lend += 7;
-		dbname[lend] = ' ';
-		lend += 1;
-		memcpy(dbname + lend, id->parameters, lenp);
-		dbname[lend + lenp] = '\0';
-		/* convert '&' to spaces */
-		for (; dbname[lend] != '\0'; lend++) {
-			if (dbname[lend] == '&' && lend > 2 &&
-					(dbname[lend-1] != '\\' || (dbname[lend-2] != '\\')))
-				dbname[lend] = ' ';
+
+		if (!use_tls) {
+			/* Handle the situation when use_tls=0 and db_url still contains tls_domain=[dom1] */
+			if (!strncmp(id->parameters, DB_TLS_DOMAIN_PARAM_EQ, DB_TLS_DOMAIN_PARAM_EQ_LEN)) {
+				char *pos = strchr(id->parameters, '&');
+				
+				unused_dom.s = id->parameters + DB_TLS_DOMAIN_PARAM_EQ_LEN;
+
+				if (!pos) {
+					unused_dom.len = strlen(id->parameters) - DB_TLS_DOMAIN_PARAM_EQ_LEN;
+				} else {
+					unused_dom.len = pos - unused_dom.s;
+				}
+			}
+			tls_domain_name = &unused_dom;
 		}
-	} else
-		dbname = id->database;
 
-	if (id->port) {
-		ports = int2str(id->port, 0);
-		LM_DBG("opening connection: postgres://xxxx:xxxx@%s:%d/%s\n", ZSW(id->host),
-			id->port, ZSW(dbname));
-		PSQL_PARAM("port", ports);
-	} else {
-		ports = NULL;
-		LM_DBG("opening connection: postgres://xxxx:xxxx@%s/%s\n", ZSW(id->host),
-			ZSW(dbname));
+		/* Set the initial dbname parameter */
+		if (tls_domain_name->len && lenp > DB_TLS_DOMAIN_PARAM_EQ_LEN+tls_domain_name->len) {
+			/* For example: tls_domain=[dom]&hostaddr=1.2.3.4&application_name=opensips */
+			/* DB_TLS_DOMAIN_PARAM_EQ_LEN = length of 'tls_donain=' (11) */
+			/* tls_domain_name->len = length of [dom] */
+			/* +1 is because the first parameter starts with a '&' and there is no need to memcpy() it */
+			/* Result is 'hostaddr=1.2.3.4 application_name=opensips' */
+			/* https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS */
+			memcpy(dbname, id->parameters+DB_TLS_DOMAIN_PARAM_EQ_LEN+tls_domain_name->len+1, lenp - (DB_TLS_DOMAIN_PARAM_EQ_LEN+tls_domain_name->len)-1);
```

Comment:
It seems that `tls_domain` should always be the first parameter. It would be nice to be able to use it in any position. If not, we should document this requirement.